### PR TITLE
[Api][Web][SIMS #639] Program and Offering Enhancements - Part 3

### PR DESCRIPTION
**Pull request by dheepak-aot**
Part 3/Final PR for #639 

**Add "Offerings" Page** (Add Study Period)
**Overview**
- [x] Change to "Add Study Period" _(Change headings/body copy that doesn’t impact devs (database etc)_
- [x] Add overview text, "Students will see the program name, description, and credential type including your study period name, dates, and year of the program."
- [x] "Study period name/offering name (e.g. 2022 Fall semester, Term 1, Cohort A)" (Max 50 Characters/ if over 59 characters product the warning, "The maximum limit is 50 characters."

**Year of Program**
- [x] Year of Program, Dropdown to select the following: "Year 1, Year 2, Year 3, Year 4, Year 5, Year 6, Year 7, Year 8,
Year 9" 
     - [x] "Display this to students (Optional)" (Present option to hide/display year to students to help identify) (Provide a pop-up of what the Students will see, "Here’s an example of what students will see."e.g. 2022 Winter co-op (Jan 15 2022 - Mar 20 2022) - Year 1"
- [x] How will this be delivered? (Onsite, Online, Blended) (If what is selected here is not aligned with what was selected in the Program, "Revise the Error Message) "Please ensure the study delivery aligns with program eligibility. To fix this, go back to your program and edit the delivery method." 

**Work-Integrated Learning** (New Section)
- [x] "Does this contain a work-integrated learning component?" (Y / N)
   - [x] If No, proceed to next section
   - [x] If Yes, Provide the following question, "What work-integrated learning type is contained within this study period?" _(Free Text)__(Add error message "This study period can’t contain a WIL component. Your program must be approved with a WIL component." (if NO in program level, can’t say yes in study period offerings page)_ 

**Study Breaks**
- [x] Add multiple study breaks functionality (including delete functionality)

 - [x] if it exceeds 10% = display warning that breaks in excess of 10% will not be funded, "This program is ineligible for StudentAid BC funding. Your break(s) can’t exceed 10% of study dates."
- [x] If breaks exceed 21 days = display error message (, "This program is ineligible for StudentAid BC funding. Your break(s) can’t exceed 21 days."
- [x] If both exceed 10% or 21 days, display error message, "This program is ineligible for StudentAid BC funding. Your break(s) can’t exceed 10% of study dates and exceed 21 days"

**Declaration** (New Section)
- [x] Provide the text, "All information is subject to verification and auditing."
- [x] "I confirm this Study Period Offering meets the policies outlined in the StudentAid BC policy manual." _(Mandatory must select)_
-------------
**Student Select Program**
**Select Program Page** 
- [x] "What capacity will you be taking this program? (Part-Time / Full Time) _Put this in a (?) A part-time program has a course load between 20 and 59%. A full-time program must have a course load of 60% or greater or 
40 and 60% for students with a permanent disability. If you are unsure, contact your school’s financial aid officer._
- [x] Year of Study
- [x] Add Courses 
- [x] Program Description _(Is now optional)

### CODE REVIEW

File: sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+COMMENT ON COLUMN sims.education_programs_offerings.year_of_study IS 'Year of study of the study period.';
+
+-- Add show_yr_of_study to education_programs_offerings table.
+ALTER TABLE
+    sims.education_programs_offerings
+ADD
+    COLUMN IF NOT EXISTS show_yr_of_study BOOLEAN;
+
+COMMENT ON COLUMN sims.education_programs_offerings.show_yr_of_study IS 'This value determins if we show year of study to students.';
```
Comment (andrewsignori-aot) [inline sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql:15]: Typo on _determins_.
Please check if the spell checker is enabled for .SQL files.

image

File: sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ADD
+    COLUMN IF NOT EXISTS year_of_study INT NOT NULL DEFAULT 1;
+
+COMMENT ON COLUMN sims.education_programs_offerings.year_of_study IS 'Year of study of the study period.';
+
+-- Add show_yr_of_study to education_programs_offerings table.
+ALTER TABLE
+    sims.education_programs_offerings
+ADD
+    COLUMN IF NOT EXISTS show_yr_of_study BOOLEAN;
```
Comment (andrewsignori-aot) [inline sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql:13]: Can it be renamed to _show_year_of_study_?

File: sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+COMMENT ON COLUMN sims.education_programs_offerings.show_yr_of_study IS 'This value determins if we show year of study to students.';
+
+-- Add has_offering_wil_component to education_programs_offerings table.
```
Comment (ann-aot) [inline sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql:17]: what is `wil` component? could you explain

File: sources/packages/api/src/utilities/eductaion-program-offering-utils.ts [hunk 4]
```diff
@@ -1,6 +1,26 @@
...
+    "Year 2",
+    "Year 3",
+    "Year 4",
+    "Year 5",
+    "Year 6",
+    "Year 7",
+    "Year 8",
+    "Year 9",
+  ];
+  return yearOfStudy[yearId - 1];
```
Comment (ann-aot) [inline sources/packages/api/src/utilities/eductaion-program-offering-utils.ts:21]: we can reduce LOC, and do something like this too ``` return `Year ${yearId}` ``` right?. just a thought

File: sources/packages/forms/src/form-definitions/educationprogram.json [hunk 5]
```diff
@@ -1655,84 +3368,242 @@
...
+      "className": "",
+      "attrs": [
+        {
+          "attr": "",
+          "value": ""
+        }
+      ],
+      "content": "Program partnerships",
+      "refreshOnChange": false,
+      "customClass": "category-header-medium primary-color has-sub-heading",
```
Comment (ann-aot) [inline sources/packages/forms/src/form-definitions/educationprogram.json:3597]: `has-sub-heading`  can we give a better name if the class

File: sources/packages/forms/src/form-definitions/educationprogram.json [hunk 6]
```diff
@@ -2857,7 +6002,243 @@
...
+          "label": "HTML",
+          "tag": "p",
+          "className": "",
+          "attrs": [
+            {
+              "attr": "",
+              "value": ""
+            }
+          ],
+          "content": "All information is subject to verification and auditing.<br>",
```
Comment (ann-aot) [inline sources/packages/forms/src/form-definitions/educationprogram.json:6096]: why `<br>` here?

File: sources/packages/forms/src/form-definitions/educationprogramoffering.json [hunk 7]
```diff
@@ -399,300 +537,380 @@
...
+          "clearOnHide": true,
+          "customDefaultValue": "",
+          "calculateValue": "",
+          "calculateServer": false,
+          "allowCalculateOverride": false,
+          "validate": {
+            "required": true,
+            "onlyAvailableItems": true,
+            "customMessage": "",
+            "custom": "valid=(data.programIntensity.includes(input))?true:'The program you have selected has only ' +data.programIntensity+ ' offering';",
```
Comment (ann-aot) [inline sources/packages/forms/src/form-definitions/educationprogramoffering.json:663]: can you add proper spacing for better code readability

File: sources/packages/forms/src/form-definitions/programinformationrequest.json [hunk 8]
```diff
@@ -1,6310 +1,7021 @@
...
-                          "calculateValue": "",
-                          "calculateServer": false,
-                          "widget": null,
-                          "attributes": {},
-                          "validateOn": "change",
-                          "validate": {
-                              "required": false,
-                              "custom": "",
-                              "customPrivate": false,
-
```
Comment (ann-aot) [inline sources/packages/forms/src/form-definitions/programinformationrequest.json]: i know this was already there, but can you pls remove theextra space before `mt-5`

Comment (ann-aot) [inline sources/packages/forms/src/form-definitions/programinformationrequest.json]: `\n<br>`?

**Review by ann-aot — COMMENTED** [review]
LGTM add some minor comments 👍

File: sources/packages/api/src/route-controllers/confirmation-of-enrollment/confirmation-of-enrollment.controller.ts [hunk 9]
```diff
@@ -172,6 +166,12 @@ export class ConfirmationOfEnrollmentController {
       ),
       applicationLocationId: application.location.id,
       applicationDeniedReason: getCOEDeniedReason(application),
+      studyBreaks: application.offering.studyBreaks?.map((studyBreak) => {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/confirmation-of-enrollment/confirmation-of-enrollment.controller.ts:169]: The return could be simplified as below.
```
application.offering.studyBreaks?.map((studyBreak) => ({
    breakStartDate: dateString(studyBreak.breakStartDate),
    breakEndDate: dateString(studyBreak.breakEndDate),
})),
```

File: sources/packages/forms/src/form-definitions/educationprogram.json [hunk 10]
```diff
@@ -1222,169 +1030,1990 @@
...
-      "values": [
+      "tree": false,
+      "components": [
+        {
+          "label": "Are students able to take this on a part time basis?",
+          "labelPosition": "top",
+          "optionsLabelPosition": "right",
+          "description": "",
+          "tooltip": " A part-time program has a course load between 20 and 59%.
A full-time program must have a course load of: 60% or greater or Between 40 and 60% for students with a permanent disability",
```
Comment (andrewsignori-aot) [inline sources/packages/forms/src/form-definitions/educationprogram.json:1081]: Please remove the white space and it seems that there is a weird character after "59%".

File: sources/packages/forms/src/form-definitions/educationprogram.json [hunk 11]
```diff
@@ -1940,42 +3770,396 @@
...
+            {
+              "label": "Join Program Error",
+              "className": "alert alert-warning fa fa-exclamation-circle",
+              "attrs": [
+                {
+                  "attr": "",
+                  "value": ""
+                }
+              ],
+              "content": "<strong> This program is ineligible for StudentAid BC funding.</strong><br/>\nAll partner institutions must be designated by StudentAid BC..",
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/forms/src/form-definitions/educationprogram.json:4040]: My understanding "\n" is formating the way as the form.io breaks the line in its editor, "<br/>" is the one that matters for the UI.

File: sources/packages/forms/src/form-definitions/educationprogramoffering.json [hunk 12]
```diff
@@ -164,59 +316,107 @@
...
           "clearOnHide": true,
           "customDefaultValue": "",
           "calculateValue": "",
           "calculateServer": false,
           "allowCalculateOverride": false,
+          "validateOn": "change",
           "validate": {
             "required": true,
-            "onlyAvailableItems": true,
+            "onlyAvailableItems": false,
```
Comment (andrewsignori-aot) [inline sources/packages/forms/src/form-definitions/educationprogramoffering.json:408]: If the list is present on form.io _onlyAvailableItems_ must be true for the proper server validation.

File: sources/packages/forms/src/form-definitions/sfaa2022-23.json [hunk 13]
```diff
@@ -2473,175 +2479,17 @@
...
-                },
                 {
                     "label": "How will you be attending the program",
                     "labelPosition": "top",
                     "widget": "choicesjs",
                     "placeholder": "Select how will you be attending the program",
                     "description": "",
-                    "tooltip": "",
+                    "tooltip": "A part-time program has a course load between 20 and 59%.
A full-time program must have a course load of 60% or greater or\r\n40 and 60% for students with a permanent disability. If you are unsure, contact your school’s financial aid officer.",
```
Comment (andrewsignori-aot) [inline sources/packages/forms/src/form-definitions/sfaa2022-23.json]: Please check the special character after "59%.".

**Review by andrewsignori-aot — COMMENTED** [review]
Great and intensive work 😉 
Please take a look at the minor comments.

File: sources/packages/forms/src/form-definitions/educationprogram.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (ann-aot) [reply sources/packages/forms/src/form-definitions/educationprogram.json:4040]: okay

**Review by guru-aot — COMMENTED** [review]
LGTM, please work on the comments.

File: sources/packages/forms/src/form-definitions/educationprogram.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/educationprogram.json:4040]: Yeah as @andrewsignori-aot mentioned next-line is used for better readability in from.io editor.
image

File: sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql (hunk 3 above)
Comment (dheepak-aot) [reply sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql:17]: WIL Abbreviated in comment.

File: sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql (hunk 1 above)
Comment (dheepak-aot) [reply sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql]: It wasn't enabled for sql files in my visual studio. I will enable it and btw. typo fixed

File: sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql]: Renaming of db column, entity, DTO and form.io done

File: sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql (hunk 3 above)
Comment (dheepak-aot) [reply sources/packages/api/src-sql/EducationProgramsOfferings/Alter-education-program-offerings.sql:17]: WIL has been abbreviated in comment.

File: sources/packages/api/src/utilities/eductaion-program-offering-utils.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/api/src/utilities/eductaion-program-offering-utils.ts]: That was good thinking. Done.

File: sources/packages/forms/src/form-definitions/educationprogram.json (hunk 5 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/educationprogram.json]: has-sub-heading removed from forms and global styles. used mt-n2 bootstrap as discussed.

File: sources/packages/forms/src/form-definitions/educationprogram.json (hunk 6 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/educationprogram.json]: br removed.

File: sources/packages/forms/src/form-definitions/educationprogramoffering.json (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/educationprogramoffering.json]: It wasn't part of my change. But added spaces for better readability.

File: sources/packages/forms/src/form-definitions/programinformationrequest.json (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/programinformationrequest.json]: removed space before mt-5

File: sources/packages/api/src/route-controllers/confirmation-of-enrollment/confirmation-of-enrollment.controller.ts (hunk 9 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/confirmation-of-enrollment/confirmation-of-enrollment.controller.ts]: Code changed.

File: sources/packages/forms/src/form-definitions/educationprogram.json (hunk 10 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/educationprogram.json]: There was a junk character when I copied the content from Zenhub. It has been removed now.

File: sources/packages/forms/src/form-definitions/educationprogramoffering.json (hunk 12 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/educationprogramoffering.json]: onlyAvailableItems is checked

File: sources/packages/forms/src/form-definitions/sfaa2022-23.json (hunk 13 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/sfaa2022-23.json]: There was a junk character when I copied the content from Zenhub. It has been removed now.

**Review by ann-aot — APPROVED** [review]
LGTM @dheepak-aot 👍 Good Work 😄

**Review by andrewsignori-aot — APPROVED** [review]
Great work, thanks for doing the changes 👍

**Review by guru-aot — APPROVED** [review]
Nice work

**Review by abschwenker — APPROVED** [review]

**Pull request merged**